### PR TITLE
allow 3rd level subtopics

### PR DESCRIPTION
**Pull request by Oluwawunmi**
### What

Refactor aggregated data subtopic pages to allow multi level subtopics
[Trello Card](https://trello.com/c/9HgeOu40/2177-refactor-aggregated-data-subtopic-pages-to-allow-3rd-level-subtopics)

### How to review

- Check code is ok
- Pull branch
- set EnableReworkedDataAggregationPages to 'true'
- run application and try accessing a route with topic, subtopic and subsubtopic in its url from your browser
- Ensure correct data is displayed and cache working.

### Who can review

Anyone

### CODE REVIEW

File: routes/routes.go [hunk 1]
```diff
@@ -41,21 +41,11 @@ func Setup(ctx context.Context, r *mux.Router, cfg *config.Config, c Clients, ca
...
-
-		// topicspecificmethodology
-		r.StrictSlash(true).Path("/{topic}/topicspecificmethodology").Methods("GET").HandlerFunc(handlers.ReadDataAggregationWithTopics(cfg, hc, cacheList, "home-methodology"))
-		r.StrictSlash(true).Path("/{topic}/{subTopic}/topicspecificmethodology").Methods("GET").HandlerFunc(handlers.ReadDataAggregationWithTopics(cfg, hc, cacheList, "home-methodology"))
+		// handle dynamic aggregated data topic pages
+		r.StrictSlash(true).Path("/{remainingPath:.*}/datalist").Methods("GET").HandlerFunc(handlers.ReadDataAggregationWithTopics(cfg, hc, cacheList, "home-datalist"))
+		r.StrictSlash(true).Path("/{remainingPath:.*}/publications").Methods("GET").HandlerFunc(handlers.ReadDataAggregationWithTopics(cfg, hc, cacheList, "home-publications"))
+		r.StrictSlash(true).Path("/{remainingPath:.*}/staticlist").Methods("GET").HandlerFunc(handlers.ReadDataAggregationWithTopics(cfg, hc, cacheList, "home-list"))
+		r.StrictSlash(true).Path("/{remainingPath:.*}/topicspecificmethodology").Methods("GET").HandlerFunc(handlers.ReadDataAggregationWithTopics(cfg, hc, cacheList, "home-methodology"))
```
Comment (lindenmckenzie) [inline routes/routes.go:48]: `remainingPath` makes it sound like it's the tail end of the path, not sure about this name.

File: cache/public/public_topic.go [hunk 2]
```diff
@@ -124,14 +102,39 @@ func UpdateDataTopics(ctx context.Context, topicClient topicCli.Clienter) func()
...
+	subtopicsChan := make(chan models.Topic)
+	dataTopicCache := getRootTopicCachePublic(ctx, subtopicsChan, topicClient, topic)
+	if dataTopicCache != nil {
+		*topics = append(*topics, dataTopicCache)
+		subTopics, err := topicClient.GetSubtopicsPublic(ctx, topicCli.Headers{}, dataTopicCache.ID)
+		if err != nil {
+			logData := log.Data{
+				"req_headers": topicCli.Headers{},
+			}
+			log.Error(ctx, "no sub topics found for "+dataTopicCache.ID+" from topic-api", err, logData)
```
Comment (lindenmckenzie) [inline cache/public/public_topic.go:115]: we try not to use dynamic log lines - better to store it in the data block.

**Review by lindenmckenzie — CHANGES_REQUESTED** [review]
Looks pretty good - have looked at code and tested some of it locally, some points:
- 3rd level unit / component testing would be good. 
- not sure about 'remainingPath'
- I'm not sure if this change has introduced this, but it's firing 1200+ requests to the topic API on startup. Given there's about 50 topics, that seems a bit heavy handed.

File: cache/private/private_topic.go [hunk 3]
```diff
@@ -67,9 +67,10 @@ func UpdateCensusTopic(ctx context.Context, serviceAuthToken string, topicClient
 func UpdateDataTopics(ctx context.Context, serviceAuthToken string, topicClient topicCli.Clienter) func() []*cache.Topic {
 	return func() []*cache.Topic {
 		var topics []*cache.Topic
+		processedTopics := make(map[string]bool)
 
-		// get the root topics subtopics (e.g. economy) from dp-topic-api
-		rootTopics, err := topicClient.GetSubtopicsPrivate(ctx, topicCli.Headers{ServiceAuthToken: serviceAuthToken}, cache.RootTopicID)
+		// get root topics from dp-topic-api (e.g. economy) from dp-topic-api
+		rootTopics, err := topicClient.GetTopicPrivate(ctx, topicCli.Headers{}, cache.RootTopicID)
```
Comment (lindenmckenzie) [inline cache/private/private_topic.go:73]: You still need to pass the ServiceAuthToken here.

Comment (lindenmckenzie) [reply cache/private/private_topic.go:73]: Recommend testing this running in publishing mode, testing against the publishing mount.

Comment (lindenmckenzie) [reply cache/private/private_topic.go:73]: I'm not sure this comment is true either anymore. Or that you retrieve rootTopics - you're retrieving the rootTopic no?

File: cache/public/public_topic.go [hunk 4]
```diff
@@ -65,52 +65,31 @@ func UpdateCensusTopic(ctx context.Context, topicClient topicCli.Clienter) func(
...
-		var rootTopicItems []models.Topic
-		if rootTopics.PublicItems != nil {
-			rootTopicItems = *rootTopics.PublicItems
+		var rootTopicItems []string
+		if rootTopics.SubtopicIds != nil {
+			rootTopicItems = *rootTopics.SubtopicIds
 		} else {
 			err := errors.New("root topic public items is nil")
 			log.Error(ctx, "failed to deference root topics items pointer", err)
 			return []*cache.Topic{cache.GetEmptyTopic()}
```
Comment (lindenmckenzie) [inline cache/public/public_topic.go:87]: This block is no longer accurate as well.

File: data/topic_filter.go [hunk 5]
```diff
@@ -34,7 +34,7 @@ type Subtopic struct {
 // Please note that only census topic filter is being returned
 func GetTopics(censusTopicCache *cache.Topic, countResp *searchModels.SearchResponse) []Topic {
 	var cachedSubtopics []cache.Subtopic
-	if censusTopicCache != nil || censusTopicCache.List != nil {
+	if censusTopicCache != nil && censusTopicCache.List != nil {
 		cachedSubtopics = censusTopicCache.List.GetSubtopics()
 	}
```
Comment (lindenmckenzie) [inline data/topic_filter.go:39]: Why this change? linting?

File: features/steps/steps.go [hunk 6]
```diff
@@ -291,15 +324,15 @@ func (c *Component) thereIsATopicAPIThatReturnsTheRootTopicAndTheSubtopicForRequ
 	c.FakeAPIRouter.topicRequest.Lock()
 	defer c.FakeAPIRouter.topicRequest.Unlock()
 
-	topicID := "6646"
+	//topicID := "6646"
```
Comment (lindenmckenzie) [inline features/steps/steps.go:327]: Why leaving the comment in?

File: handlers/handlers_test.go [hunk 7]
```diff
@@ -669,6 +671,82 @@ func TestUnitReadDataAggregationWithTopicsSuccess(t *testing.T) {
...
+				secondSearchCall := mockedSearchClient.GetSearchCalls()[1]
+
+				firstCallTopic := firstSearchCall.Options.Query.Get("topics")
+				secondCallTopic := secondSearchCall.Options.Query.Get("topics")
+
+				if firstCallTopic != "" {
+					So(firstCallTopic, ShouldEqual, testSubSubTopic.ID)
+				} else {
+					So(secondCallTopic, ShouldEqual, testSubSubTopic.ID)
+				}
```
Comment (lindenmckenzie) [inline handlers/handlers_test.go:746]: I think I put in a function that deals with getting the right search call?

File: routes/routes.go [hunk 8]
```diff
@@ -41,21 +41,13 @@ func Setup(ctx context.Context, r *mux.Router, cfg *config.Config, c Clients, ca
...
-		r.StrictSlash(true).Path("/{topic}/{subTopic}/publications").Methods("GET").HandlerFunc(handlers.ReadDataAggregationWithTopics(cfg, hc, cacheList, "home-publications"))
-
-		// staticlist
-		r.StrictSlash(true).Path("/{topic}/staticlist").Methods("GET").HandlerFunc(handlers.ReadDataAggregationWithTopics(cfg, hc, cacheList, "home-list"))
-		r.StrictSlash(true).Path("/{topic}/{subTopic}/staticlist").Methods("GET").HandlerFunc(handlers.ReadDataAggregationWithTopics(cfg, hc, cacheList, "home-list"))
-
-		// topicspecificmethodology
-		r.StrictSlash(true).Path("/{topic}/topicspecificmethodology").Methods("GET").HandlerFunc(handlers.ReadDataAggregationWithTopics(cfg, hc, cacheList, "home-methodology"))
-		r.StrictSlash(true).Path("/{topic}/{subTopic}/topicspecificmethodology").Methods("GET").HandlerFunc(handlers.ReadDataAggregationWithTopics(cfg, hc, cacheList, "home-methodology"))
+		if cfg.EnableTopicAggregationPages {
```
Comment (lindenmckenzie) [inline routes/routes.go:44]: If this is a combination feature flag then please note this in the readme.

File: service/service.go [hunk 9]
```diff
@@ -99,11 +99,14 @@ func (svc *Service) Init(ctx context.Context, cfg *config.Config, serviceList *E
 
 	if cfg.IsPublishing {
 		svc.Cache.CensusTopic.AddUpdateFunc(cache.CensusTopicID, cachePrivate.UpdateCensusTopic(ctx, cfg.ServiceAuthToken, clients.Topic))
-		svc.Cache.DataTopic.AddUpdateFuncs(cachePrivate.UpdateDataTopics(ctx, cfg.ServiceAuthToken, clients.Topic))
+		if svc.Config.EnableTopicAggregationPages {
```
Comment (lindenmckenzie) [inline service/service.go:102]: why svc.Config and not cfg. direct?

File: service/service.go [hunk 10]
```diff
@@ -99,11 +99,14 @@ func (svc *Service) Init(ctx context.Context, cfg *config.Config, serviceList *E
...
+			svc.Cache.DataTopic.AddUpdateFuncs(cachePrivate.UpdateDataTopics(ctx, cfg.ServiceAuthToken, clients.Topic))
+		}
 	} else {
 		svc.Cache.CensusTopic.AddUpdateFunc(cache.CensusTopicID, cachePublic.UpdateCensusTopic(ctx, clients.Topic))
-		svc.Cache.DataTopic.AddUpdateFuncs(cachePublic.UpdateDataTopics(ctx, clients.Topic))
-
+		if svc.Config.EnableTopicAggregationPages {
```
Comment (lindenmckenzie) [inline service/service.go:107]: Same question

File: service/service.go [hunk 11]
```diff
@@ -140,7 +143,9 @@ func (svc *Service) Run(ctx context.Context, svcErrors chan error) {
 
 	// Start caching
 	go svc.Cache.CensusTopic.StartUpdates(ctx, svcErrors)
-	go svc.Cache.DataTopic.StartUpdates(ctx, svcErrors)
+	if svc.Config.EnableTopicAggregationPages {
```
Comment (lindenmckenzie) [inline service/service.go:146]: Same question.

File: service/service.go [hunk 12]
```diff
@@ -168,7 +173,9 @@ func (svc *Service) Close(ctx context.Context) error {
 
 		// stop caching
 		svc.Cache.CensusTopic.Close()
-		svc.Cache.DataTopic.Close()
+		if svc.Config.EnableTopicAggregationPages {
```
Comment (lindenmckenzie) [inline service/service.go:176]: Same question.

File: config/config.go [hunk 13]
```diff
@@ -25,7 +25,8 @@ type Config struct {
 	DefaultAggregationSort                  string        `envconfig:"DEFAULT_AGGREGATION_SORT"`
 	DefaultSort                             string        `envconfig:"DEFAULT_SORT"`
 	DefaultDatasetSort                      string        `envconfig:"DEFAULT_DATASET_SORT"`
-	EnableReworkedDataAggregationPages      bool          `envconfig:"ENABLE_REWORKED_DATA_AGGREGATION_PAGES"`
+	EnableAggregationPages                  bool          `envconfig:"ENABLE_AGGREGATION_PAGES"`
+	EnableTopicAggregationPages             bool          `envconfig:"ENABLE_TOPIC_AGGREGATION_PAGES"`
```
Comment (lindenmckenzie) [inline config/config.go:29]: There should be units tests for this.

File: service/service.go (hunk 9 above)
Comment (lindenmckenzie) [inline service/service.go:102]: I would suggest a component level test to ensure no requests go to the Topic API if it's switched off. Or I guess maybe only to the root topic. Something to think about. Not dead set on it.

**Review by lindenmckenzie — CHANGES_REQUESTED** [review]

File: README.md [hunk 14]
```diff
@@ -44,7 +44,8 @@ No further dependencies other than those defined in `go.mod`
 | DEFAULT_OFFSET                              | 0                               | The default offset of search results                                                                               |
 | DEFAULT_PAGE                                | 1                               | The default current page of search results                                                                         |
 | DEFAULT_SORT                                | relevance                       | The default sort of search results                                                                                 |
-| ENABLE_REWORKED_DATA_AGGREGATION_PAGES      | false                           | Enable the reworked data aggregation pages                                                                         |
+| ENABLE_AGGREGATION_PAGES                    | false                           | Enable the aggregation pages                                                                                       |
```
[the remark this answers is not in the document]
Comment (Oluwawunmi) [reply README.md:47]: The plan is add the new flags first all the way up to prod, then come back to delete the old flag all the way up after deployment.

File: cache/private/private_topic.go [hunk 15]
```diff
@@ -120,19 +100,60 @@ func UpdateDataTopics(ctx context.Context, serviceAuthToken string, topicClient
...
+	// Get the topic details from the topic client
+	dataTopic, err := topicClient.GetTopicPrivate(ctx, topicCli.Headers{ServiceAuthToken: serviceAuthToken}, topicID)
+	if err != nil {
+		log.Error(ctx, "failed to get topic details from topic-api", err)
+		return
+	}
+
+	if dataTopic != nil {
+		// Append the current topic to the list of topics
+		// subtopicsChan := make(chan models.Topic)
```
Comment (lindenmckenzie) [inline cache/private/private_topic.go:122]: Why commented out?

File: cache/public/public_topic.go [hunk 16]
```diff
@@ -119,19 +98,61 @@ func UpdateDataTopics(ctx context.Context, topicClient topicCli.Clienter) func()
...
+	// Get the topic details from the topic client
+	dataTopic, err := topicClient.GetTopicPublic(ctx, topicCli.Headers{}, topicID)
+	if err != nil {
+		log.Error(ctx, "failed to get topic details from topic-api", err)
+		return
+	}
+
+	if dataTopic != nil {
+		// Append the current topic to the list of topics
+		// subtopicsChan := make(chan models.Topic)
```
Comment (lindenmckenzie) [inline cache/public/public_topic.go:120]: Why commented out?

File: README.md [hunk 17]
```diff
@@ -44,7 +44,8 @@ No further dependencies other than those defined in `go.mod`
 | DEFAULT_OFFSET                              | 0                               | The default offset of search results                                                                               |
 | DEFAULT_PAGE                                | 1                               | The default current page of search results                                                                         |
 | DEFAULT_SORT                                | relevance                       | The default sort of search results                                                                                 |
-| ENABLE_REWORKED_DATA_AGGREGATION_PAGES      | false                           | Enable the reworked data aggregation pages                                                                         |
+| ENABLE_AGGREGATION_PAGES                    | false                           | Enable the aggregation pages, is a combination feature flag with ENABLE_TOPIC_AGGREGATION_PAGES                    |
+| ENABLE_TOPIC_AGGREGATION_PAGES              | false                           | Enable the topic aggregation pages, is a combination feature flag with ENABLE_AGGREGATION_PAGES                    |
```
Comment (lindenmckenzie) [inline README.md:48]: So just be clear here that to enable the topic ones you have to enable the aggregation pages as well.

**Review by lindenmckenzie — COMMENTED** [review]
Exceptionally pedantic comments.

**Review by lindenmckenzie — APPROVED** [review]

**Review by lindenmckenzie — APPROVED** [review]

**Pull request merged**